Redis: initial implementation.

jelmersnoeck opened the pull request:
Need to fix several things up. Example: DeleteMulti should do 1 Redis call.

jelmersnoeck committed:
MemoryTest: move to internal tester package.

This way we can reuse these tests for other interfaces.

jelmersnoeck committed:
Redis: add initial implementation.

jelmersnoeck committed:
Travis: add Redis.

matthiasmullie commented on redis.go:
```diff
@@ -0,0 +1,190 @@
+// there is a value present, we will add the given offset to that value and
+// update the value with the new TTL.
+func (c *RedisCache) incrementOffset(key string, initial, offset, ttl int64) bool {
+	if !c.exists(key) {
```
Hier hebde een mogelijks concurrency probleem: dezen is niet atomic.

Ge gaat eerst checken of de key exists. Zoniet gaat ge een initial value setten. Mogelijks probleem: een ander proces kan tussen die 2 operaties al een andere waarde opgeslagen hebben.
Later gaat ge dan nog een get &INCRBY doen. Maar ondertussen kan ook alweer een ander proces andere dingen naar die key geschreven hebben.

Enfin, als ge meerdere operaties moet doen moet ge kunnen garanderen dat er extern niets de data kan veranderen ondertussen. Redis heeft transactions (http://redis.io/commands/multi): ge kunt eerst een transaction starten, dan lezen, nieuwe value opslaan, en dan pas in 1 keer committen (http://redis.io/commands/exec)

matthiasmullie commented on redis.go:
```diff
@@ -0,0 +1,190 @@
+		return false
+	}
+
+	_, err := c.client.Do("INCRBY", key, offset)
```
Ik zou geen INCRBY gebruiken: ge kunt immers uw TTL niet instellen.
In plaats daarvan doe ik de increment zelf in code (ge hebt immers toch al de value opgevraagd) en dan set ik die nieuwe value (mét TTL), binnen 1 transactie.

jelmersnoeck committed:
Bytes: move to internal numbers package.

jelmersnoeck committed:
Tests: use test driver table.

jelmersnoeck committed:
CacherTest: redundant test.

We create an array of cacher.Cacher instances. If the implementation
does not justify the interface, we'll see an error there.

jelmersnoeck committed:
NumbersTest: add copyright.

jelmersnoeck committed:
Redis: use `Set` for increments.

Redis' `INCRBY` doesn't support TTL. We'll manually increment this
information.

jelmersnoeck committed:
Redis: use transactions for increments.

jelmersnoeck committed:
Redis: use MGET to get multiple keys.

jelmersnoeck committed:
Adapters: negative values for TTL.

jelmersnoeck committed:
Redis: optimized DeleteMulti.

jelmersnoeck committed:
Redis: add delete multi transactions.

jelmersnoeck committed:
Adapters: make own packages.

jelmersnoeck committed:
Readme: add memory example.

jelmersnoeck committed:
Readme: add Redis documentation.

jelmersnoeck committed:
Redis: transactions for replace.
